src/csv/lexer.cr

```diff
@@ -89,15 +89,15 @@ abstract class CSV::Lexer
       when '\0'
         raise "unclosed quote"
         break
-      when '"'
+      when @quote_char
         case next_char
-        when ','
+        when @separator
           check_last_empty_column
           break
         when '\r', '\n', '\0'
           break
-        when '"'
-          @buffer << '"'
+        when @quote_char
```

Comment:
@luislavena The instance variable might change between the if's condition and its body, if there's a Thread context switch (which we don't have right now, but we will). So this code will behave in a wrong way if that happens, but at most an exception will be raised on badly parsed input or something like that.

However, if we allow things like this to work:

```crystal
# @something : Int32 | Nil
if @something
  # Now compiler assumes @something is Int32
  @something + 2
end
```

and there's a Thread context switch in the middle and `@something` becomes `nil`, now the `+` method doesn't make sense anymore and the runtime will execute garbage code and the program will crash in a horrible way, or maybe not and there will be undefined behavior, stack corruption, etc.

So the compiler is only preventing undefined behaviour/crashes, but not data races.